<?xml version="1.0" ?>
<launch>
  <arg name="robot_ip" default="172.16.0.2"/>
  <arg name="load_gripper" default="true" />
  <arg name="arm_id" default="panda_leader" />

    <group ns="$(arg arm_id)">
        
        <param name="robot_description" command="$(find xacro)/xacro $(find panda_teleoperation)/xacro/franka_arm.urdf.xacro arm_id:=$(arg arm_id) hand:=$(arg load_gripper)" />

        <include file="$(find franka_gripper)/launch/franka_gripper.launch" if="$(arg load_gripper)">
            <arg name="robot_ip" value="$(arg robot_ip)" />
        </include>

        <node name="franka_control" pkg="franka_control" type="franka_control_node" output="screen" required="true">
            <rosparam command="load" file="$(find panda_teleoperation)/config/panda_leader_control_node.yaml" />
            <param name="robot_ip" value="$(arg robot_ip)" />
        </node>

        <rosparam command="load" file="$(find panda_teleoperation)/config/panda_leader_controller.yaml" />
        <node name="state_controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="panda_leader_state_controller"/>
        <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen"/>
   
        <node name="joint_state_publisher" type="joint_state_publisher" pkg="joint_state_publisher" output="screen">
            <rosparam if="$(arg load_gripper)" param="source_list">[panda_leader_state_controller/joint_states, franka_gripper/joint_states] </rosparam>
            <rosparam unless="$(arg load_gripper)" param="source_list">[panda_leader_state_controller/joint_states] </rosparam>
            <param name="rate" value="30"/>
        </node>
    
        <node name="joint_state_desired_publisher" type="joint_state_publisher" pkg="joint_state_publisher" output="screen">
            <rosparam if="$(arg load_gripper)" param="source_list">[panda_leader_state_controller/joint_states_desired, franka_gripper/joint_states] </rosparam>
            <rosparam unless="$(arg load_gripper)" param="source_list">[panda_leader_state_controller/joint_states_desired] </rosparam>
            <param name="rate" value="30"/>
            <remap from="/joint_states" to="/joint_states_desired" />
        </node> 

        <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="panda_leader_cartesian_impedance_controller"/>
        <!-- <node name="rqt_reconfigure" pkg="rqt_reconfigure" type="rqt_reconfigure" required="false" /> -->


    </group>

</launch>